### PR TITLE
align 2.11 slurm files with 3.x

### PR DESCRIPTION
**Pull request by Nbot07A**
### Description of changes
*align the compute and head node files between 2.11 and 3.x

### References
* Link to documentation useful to understand the changes.
Minor fix to slurm service files: https://github.com/aws/aws-parallelcluster-cookbook/commit/6fd9f377186f53a4d22eeb0df1e29585f9b2f5ce
Update slurmd.service and slurmctld.service
: https://github.com/aws/aws-parallelcluster-cookbook/commit/d264c6b9c1b751eac761bf48aaff88ca0e794cdf
Factor out slurm installation dir: https://github.com/aws/aws-parallelcluster-cookbook/commit/def877110c2bf1562be58b93e1f5550f469411b7#diff-5c42e5f029215133e088c66f03fb437d02e7605c303046b6694acce4bb70f8db

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: files/default/slurmctld.service [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
-After=network.target munge.service
+After=network-online.target munge.service
+Wants=network-online.target
 ConditionPathExists=/opt/slurm/etc/slurm.conf
 
 [Service]
 Type=simple
 EnvironmentFile=-/etc/sysconfig/slurmctld
-ExecStart=/opt/slurm/sbin/slurmctld -D $SLURMCTLD_OPTIONS
+ExecStart=<%= node['cluster']['slurm']['install_dir'] %>/sbin/slurmctld -D -s $SLURMCTLD_OPTIONS
```
Comment (chenwany) [inline files/default/slurmctld.service]: In 3.x, we use `default['cluster']['slurm']['install_dir'] = "/opt/slurm"`, so it is `node['cluster']['slurm']['install_dir']` here, but in 2.11, the node attribute is `default['cfncluster']['slurm']['install_dir'] = "/opt/slurm"`, we need to use `node['cfncluster']['slurm']['install_dir']` instead.

Comment (chenwany) [reply files/default/slurmctld.service]: This file need to be renamed to be end with .erb and then put in `https://github.com/aws/aws-parallelcluster-cookbook/tree/release-2.11-slurm2108/templates/default/slurm`

File: files/default/slurmd.service [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
-After=munge.service network.target remote-fs.target
+After=munge.service network-online.target remote-fs.target
+Wants=network-online.target
 ConditionPathExists=/opt/slurm/etc/slurm.conf
 
 [Service]
 Type=simple
 EnvironmentFile=-/etc/sysconfig/slurmd
-ExecStart=/opt/slurm/sbin/slurmd -D $SLURMD_OPTIONS
+ExecStart=<%= node['cluster']['slurm']['install_dir'] %>/sbin/slurmd -D -s $SLURMD_OPTIONS
```
Comment (chenwany) [inline files/default/slurmd.service]: This file need to be renamed to be end with.erb and then put in https://github.com/aws/aws-parallelcluster-cookbook/tree/release-2.11-slurm2108/templates/default/slurm
You can put it under here: `slurm/compute/slurmd.service.erb`

File: recipes/compute_slurm_config.rb [hunk 3]
```diff
@@ -26,8 +26,8 @@
 
 # Mount /opt/slurm over NFS
 # Computemgtd config is under /opt/slurm/etc/pcluster; all compute nodes share a config
-mount '/opt/slurm' do
-  device(lazy { "#{node['cfncluster']['cfn_master_private_ip']}:/opt/slurm" })
+mount "#{node['cluster']['slurm']['install_dir']}" do
```
Comment (chenwany) [inline recipes/compute_slurm_config.rb]: `node['cluster']` should be `node['cfncluster']`, same for other places

File: templates/default/slurm/slurm.conf.erb [hunk 4]
```diff
@@ -32,6 +32,7 @@ CommunicationParameters=NoAddrCache
 SuspendProgram=<%= node['cfncluster']['scripts_dir'] %>/slurm/slurm_suspend
 ResumeProgram=<%= node['cfncluster']['scripts_dir'] %>/slurm/slurm_resume
 ResumeFailProgram=<%= node['cfncluster']['scripts_dir'] %>/slurm/slurm_suspend
+RebootProgram=/sbin/reboot
```
Comment (chenwany) [inline templates/default/slurm/slurm.conf.erb]: Isn't the change related to `RebootProgram=/sbin/reboot` already merge?

Comment (Nbot07A) [reply templates/default/slurm/slurm.conf.erb]: Yes

File: recipes/compute_slurm_config.rb (hunk 3 above)
Comment (Nbot07A) [reply recipes/compute_slurm_config.rb]: ok, I'll fix that now

File: files/default/slurmd.service (hunk 2 above)
Comment (Nbot07A) [reply files/default/slurmd.service]: ok, will do

File: files/default/slurmctld.service (hunk 1 above)
Comment (Nbot07A) [reply files/default/slurmctld.service]: ok, I'll change that now

File: recipes/compute_slurm_config.rb [hunk 5]
```diff
@@ -26,8 +26,8 @@
 
 # Mount /opt/slurm over NFS
 # Computemgtd config is under /opt/slurm/etc/pcluster; all compute nodes share a config
-mount '/opt/slurm' do
-  device(lazy { "#{node['cfncluster']['cfn_master_private_ip']}:/opt/slurm" })
+mount "#{node['cfncluster']['slurm']['install_dir']}" do
+  device(lazy { "#{node['cfncluster']['head_node_private_ip']}:#{node['cfncluster']['slurm']['install_dir']}" })
```
Comment (chenwany) [inline recipes/compute_slurm_config.rb]: `head_node_private_ip` should be `cfn_master_private_ip` in 2.11, we don't have attribute `head_node_private_ip`. You can refer to the variable specified in `attributes/default.rb`

File: templates/default/slurm/slurmd.service.erb [hunk 6]
```diff
@@ -1,12 +1,13 @@
...
-After=munge.service network.target remote-fs.target
+After=munge.service network-online.target remote-fs.target
+Wants=network-online.target
 ConditionPathExists=/opt/slurm/etc/slurm.conf
 
 [Service]
 Type=simple
 EnvironmentFile=-/etc/sysconfig/slurmd
-ExecStart=/opt/slurm/sbin/slurmd -D $SLURMD_OPTIONS
+ExecStart=<%= node['cfncluster']['slurm']['install_dir'] %>/sbin/slurmd -D -s $SLURMD_OPTIONS
```
Comment (chenwany) [inline templates/default/slurm/slurmd.service.erb]: Please create a folder `compute_node` and move the file under /slurm/compute/ as you have:
```
template '/etc/systemd/system/slurmd.service' do
  source 'slurm/compute/slurmd.service.erb'
  owner 'root'
  group 'root'
  mode '0644'
  action :create
  only_if { node['init_package'] == 'systemd' }
end
```
in the code

File: templates/default/slurm/slurmctld.service.erb [hunk 7]
```diff
@@ -1,12 +1,13 @@
 [Unit]
 Description=Slurm controller daemon
-After=network.target munge.service
+After=network-online.target munge.service
```
Comment (chenwany) [inline templates/default/slurm/slurmctld.service.erb]: Please create a folder head_node and move the file under /slurm/head_node/ as you have:
```
template "#{node['cfncluster']['slurm']['install_dir']}/etc/slurm.csh" do
  source 'slurm/head_node/slurm.csh.erb'
  owner 'root'
  group 'root'
  mode '0755'
end
```
in the code

File: recipes/head_node_slurm_config.rb [hunk 8]
```diff
@@ -91,26 +91,26 @@
...
+  source 'slurm/head_node/slurm.sh.erb'
   owner 'root'
   group 'root'
   mode '0755'
 end
 
-cookbook_file '/opt/slurm/etc/slurm.csh' do
-  source 'slurm.csh'
+template "#{node['cfncluster']['slurm']['install_dir']}/etc/slurm.csh" do
+  source 'slurm/head_node/slurm.csh.erb'
```
Comment (chenwany) [inline recipes/head_node_slurm_config.rb:113]: Since this is change, we need to have the change include `cookbooks/aws-parallelcluster-slurm/templates/default/slurm/head_node/slurm.csh.erb` and remove `cookbooks/aws-parallelcluster-slurm/files/default/head_node_slurm/slurm.csh`, same for `'slurm/head_node/slurm.sh.erb'`

**Review by chenwany — COMMENTED** [review]
Could you update the change log to add `- Update slurmctld and slurmd systemd service files.`?
Could you change the PR description?

File: templates/default/slurm/slurmd.service.erb (hunk 6 above)
Comment (Nbot07A) [reply templates/default/slurm/slurmd.service.erb]: I just wanted to clarify. Since the template states the /slurm/compute directory. Would that mean that I need to move the file slurmd.service to, and create if necessary, the /slurm/compute directory? Also, why would I create a compute_node folder?

Comment (chenwany) [reply templates/default/slurm/slurmd.service.erb]: Yes, please create a `compute` folder. My bad to refer it to `compute_node`

File: recipes/compute_slurm_config.rb [hunk 9]
```diff
@@ -26,8 +26,8 @@
 
 # Mount /opt/slurm over NFS
 # Computemgtd config is under /opt/slurm/etc/pcluster; all compute nodes share a config
-mount '/opt/slurm' do
-  device(lazy { "#{node['cfncluster']['cfn_master_private_ip']}:/opt/slurm" })
+mount (node['cfncluster']['slurm']['install_dir']).to_s do
```
Comment (chenwany) [inline recipes/compute_slurm_config.rb]: Could you align with 3.x here instead of using ().to_s here?

File: recipes/head_node_slurm_finalize.rb [hunk 10]
```diff
@@ -49,14 +49,14 @@
     require 'shellwords'
 
     # Example output for sinfo
-    # $ /opt/slurm/bin/sinfo -N -h -o '%N %t'
+    # $ #{node['cfncluster']['slurm']['install_dir']}/bin/sinfo -N -h -o '%N %t'
```
Comment (chenwany) [inline recipes/head_node_slurm_finalize.rb:52]: minor: I think we don't need to change it in the comment. Because in the comment we are running linux command, which is not related to chef node attribute.

File: recipes/head_node_slurm_finalize.rb [hunk 11]
```diff
@@ -71,15 +71,15 @@
     require 'chef/mixin/shell_out'
     require 'shellwords'
 
-    # $ /opt/slurm/bin/squeue --name='parallelcluster-init-cluster' -O state -h
+    # $ #{node['cfncluster']['slurm']['install_dir']}/bin/squeue --name='parallelcluster-init-cluster' -O state -h
```
Comment (chenwany) [inline recipes/head_node_slurm_finalize.rb:74]: minor: I think we don't need to change it in the comment. Because in the comment we are running linux command, which is not related to chef node attribute.

File: templates/default/slurm/compute/slurmd.service.erb [hunk 12]
```diff
@@ -1,12 +1,13 @@
 [Unit]
 Description=Slurm node daemon
-After=munge.service network.target remote-fs.target
+After=munge.service network-online.target remote-fs.target
+Wants=network-online.target
 ConditionPathExists=/opt/slurm/etc/slurm.conf
```
Comment (chenwany) [inline templates/default/slurm/compute/slurmd.service.erb]: Missing change of /opt/slurm/ here

File: templates/default/slurm/head_node/slurmctld.service.erb [hunk 13]
```diff
@@ -1,12 +1,13 @@
 [Unit]
 Description=Slurm controller daemon
-After=network.target munge.service
+After=network-online.target munge.service
+Wants=network-online.target
 ConditionPathExists=/opt/slurm/etc/slurm.conf
```
Comment (chenwany) [inline templates/default/slurm/head_node/slurmctld.service.erb]: Missing change of /opt/slurm here

File: libraries/helpers.rb [hunk 14]
```diff
@@ -204,11 +204,11 @@ def hit_dynamodb_info
   require 'chef/mixin/shell_out'
 
   output = shell_out!("#{node['cfncluster']['cookbook_virtualenv_path']}/bin/aws dynamodb " \
-    "--region #{node['cfncluster']['cfn_region']} query --table-name #{node['cfncluster']['cfn_ddb_table']} " \
-    "--index-name InstanceId --key-condition-expression 'InstanceId = :instanceid' " \
-    "--expression-attribute-values '{\":instanceid\": {\"S\":\"#{node['ec2']['instance_id']}\"}}' " \
-    "--projection-expression 'Id,MasterPrivateIp,MasterHostname' " \
-    "--output text --query 'Items[0].[Id.S,MasterPrivateIp.S,MasterHostname.S]'", user: 'root').stdout.strip
+                      "--region #{node['cfncluster']['cfn_region']} query --table-name #{node['cfncluster']['cfn_ddb_table']} " \
```
Comment (chenwany) [inline libraries/helpers.rb]: these changes are not related to this PR

File: recipes/arm_pl_install.rb [hunk 15]
```diff
@@ -17,7 +17,7 @@
 
 return unless node['conditions']['arm_pl_supported']
 
-armpl_installer = "#{node['cfncluster']['sources_dir']}/"\
+armpl_installer = "#{node['cfncluster']['sources_dir']}/" \
```
Comment (chenwany) [inline recipes/arm_pl_install.rb]: Change is not related to this PR

File: recipes/cloudwatch_agent_config.rb [hunk 16]
```diff
@@ -76,8 +76,8 @@
     'CONFIG_DATA_PATH' => config_data_path
   )
   not_if { ::File.exist?('/opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.json') }
-  command "#{node.default['cfncluster']['cookbook_virtualenv_path']}/bin/python #{config_script_path} "\
-          "--platform #{node['platform']} --config $CONFIG_DATA_PATH --log-group $LOG_GROUP_NAME "\
+  command "#{node.default['cfncluster']['cookbook_virtualenv_path']}/bin/python #{config_script_path} " \
```
Comment (chenwany) [inline recipes/cloudwatch_agent_config.rb]: change that is not related to this PR

File: recipes/dcv_config.rb [hunk 17]
```diff
@@ -106,9 +106,9 @@ def allow_gpu_acceleration
     # * user to make owner of the two files
     # * group to make owner of the two files
     # NOTE: the last arg is hardcoded to be 'dcv' so that the dcvserver can read the files when authenticating
-    command "/etc/parallelcluster/generate_certificate.sh"\
-            " \"#{node['cfncluster']['dcv']['authenticator']['certificate']}\""\
-            " \"#{node['cfncluster']['dcv']['authenticator']['private_key']}\""\
+    command "/etc/parallelcluster/generate_certificate.sh" \
```
Comment (chenwany) [inline recipes/dcv_config.rb]: Change not related to this PR

File: recipes/nvidia_install.rb [hunk 18]
```diff
@@ -91,7 +91,7 @@
   if node['platform'] == 'ubuntu'
     # For ubuntu, CINC17 apt-package resources need full versions for `version`
     execute "install_fabricmanager_for_ubuntu" do
-      command "apt -y install #{node['cfncluster']['nvidia']['fabricmanager']['package']}=#{node['cfncluster']['nvidia']['fabricmanager']['version']} "\
+      command "apt -y install #{node['cfncluster']['nvidia']['fabricmanager']['package']}=#{node['cfncluster']['nvidia']['fabricmanager']['version']} " \
```
Comment (chenwany) [inline recipes/nvidia_install.rb]: the change is not related to this PR

File: recipes/ganglia_config.rb [hunk 19]
```diff
@@ -62,7 +62,7 @@
 
   # For ComputeFleet and MasterServer
 
-  if node['platform'] == 'centos' && node['platform_version'].to_i >= 7 || node['platform'] == 'amazon'
+  if (node['platform'] == 'centos' && node['platform_version'].to_i >= 7) || node['platform'] == 'amazon'
```
Comment (chenwany) [inline recipes/ganglia_config.rb]: The change is not related to this PR

**Review by chenwany — COMMENTED** [review]
there are changes related to /opt/slurm missing in 
templates/default/slurm/head_node/slurm.csh.erb
and
templates/default/slurm/head_node/slurm.sh.erb

File: .rubocop.yml [hunk 20]
```diff
@@ -61,7 +59,7 @@ Style/StringChars: # (new in 1.12)
 Style/SwapValues: # (new in 1.1)
   Enabled: true
 
-Gemspec/DateAssignment: # (new in 1.10)
+Gemspec/DeprecatedAttributeAssignment: # (new in 1.10)
```
Comment (chenwany) [inline .rubocop.yml]: Why this line is change? Could you put changes related to rubocop configuration on a single commit or a separate PR without including slurm change? (Or we can simply merge this PR first: https://github.com/aws/aws-parallelcluster-cookbook/pull/1571)

File: recipes/compute_slurm_config.rb (hunk 9 above)
Comment (Nbot07A) [reply recipes/compute_slurm_config.rb]: Yes

Comment (Nbot07A) [reply recipes/compute_slurm_config.rb]: Actually, sorry, it seems this file is different from what is depicted here.

Comment (Nbot07A) [reply recipes/compute_slurm_config.rb]: Sorry again, this file is as depicted and I'm pushing the change now

File: .rubocop.yml (hunk 20 above)
Comment (Nbot07A) [reply .rubocop.yml]: Sorry, I thought I never staged this file. I made this change following my local removal of gitlab-security and encountering an error that suggested this change.

File: recipes/update_head_node_slurm.rb [hunk 21]
```diff
@@ -16,17 +16,17 @@
...
+                       " --key #{node['cfncluster']['cluster_config_s3_key']}" \
                        " --region #{node['cfncluster']['cfn_region']} #{updated_cluster_config_path}"
 fetch_config_command += " --version-id #{node['cfncluster']['cluster_config_version']}" unless node['cfncluster']['cluster_config_version'].nil?
 shell_out!(fetch_config_command)
 if !File.exist?(node['cfncluster']['cluster_config_path']) || !FileUtils.identical?(updated_cluster_config_path, node['cfncluster']['cluster_config_path'])
   # Generate pcluster specific configs
   execute "generate_pcluster_slurm_configs" do
     command "#{node['cfncluster']['cookbook_virtualenv_path']}/bin/python #{node['cfncluster']['scripts_dir']}/slurm/pcluster_slurm_config_generator.py" \
-            " --output-directory /opt/slurm/etc/ --template-directory #{node['cfncluster']['scripts_dir']}/slurm/templates/ --input-file #{updated_cluster_config_path}"
+            " --output-directory #{node['cfncluster']['slurm']['install_dir']}/etc/ --template-directory #{node['cfncluster']['scripts_dir']}/slurm/templates/ --input-file #{updated_cluster_config_path}"
```
Comment (chenwany) [inline recipes/update_head_node_slurm.rb:29]: This line is too long: Layout/LineLength: Line is too long. [203/200]
Can you break it into two lines with  `\`?
```
" --output-directory #{node['cfncluster']['slurm']['install_dir']}/etc/ --template-directory #{node['cfncluster']['scripts_dir']}/slurm/templates/ --input-file #{updated_cluster_config_path}"
```

File: recipes/head_node_slurm_finalize.rb [hunk 22]
```diff
@@ -49,14 +49,14 @@
...
-    # $ /opt/slurm/bin/sinfo -N -h -o '%N %t'
+    # $ #{node['cfncluster']['slurm']['install_dir']}/bin/sinfo -N -h -o '%N %t'
     # ondemand-dy-c5.2xlarge-1 idle~
     # ondemand-dy-c5.2xlarge-2 idle~
     # spot-dy-c5.xlarge-1 idle~
     # spot-st-t2.large-1 down
     # spot-st-t2.large-2 idle
     is_fleet_ready_command = Shellwords.escape(
-      "set -o pipefail && /opt/slurm/bin/sinfo -N -h -o '%N %t' | { grep -E '^[a-z0-9\\-]+\\-st\\-[a-z0-9]+\\-[0-9]+ .*' || true; } | { grep -v -E '(idle|alloc|mix)$' || true; }"
+      "set -o pipefail && #{node['cfncluster']['slurm']['install_dir']}/bin/sinfo -N -h -o '%N %t' | { grep -E '^[a-z0-9\\-]+\\-st\\-[a-z0-9]+\\-[0-9]+ .*' || true; } | { grep -v -E '(idle|alloc|mix)$' || true; }"
```
Comment (chenwany) [inline recipes/head_node_slurm_finalize.rb:59]: This line is too long: C: Layout/LineLength: Line is too long. [213/200]
Could you break it into two lines with `\`?

File: recipes/head_node_slurm_finalize.rb [hunk 23]
```diff
@@ -71,15 +71,15 @@
     require 'chef/mixin/shell_out'
     require 'shellwords'
 
-    # $ /opt/slurm/bin/squeue --name='parallelcluster-init-cluster' -O state -h
+    # $ #{node['cfncluster']['slurm']['install_dir']}/bin/squeue --name='parallelcluster-init-cluster' -O state -h
     # CONFIGURING
     # RUNNING
-    are_jobs_running_command = Shellwords.escape("set -o pipefail && /opt/slurm/bin/squeue --name='parallelcluster-init-cluster' -O state -h | { grep -v 'RUNNING' || true; }")
+    are_jobs_running_command = Shellwords.escape("set -o pipefail && #{node['cfncluster']['slurm']['install_dir']}/bin/squeue --name='parallelcluster-init-cluster' -O state -h | { grep -v 'RUNNING' || true; }")
```
Comment (chenwany) [inline recipes/head_node_slurm_finalize.rb:77]: This line is too long: C: Layout/LineLength: Line is too long. [210/200]
Could you break it into two lines with `\`?

**Comment by Nbot07A** [comment]
See new PR https://github.com/aws/aws-parallelcluster-cookbook/pull/1573
